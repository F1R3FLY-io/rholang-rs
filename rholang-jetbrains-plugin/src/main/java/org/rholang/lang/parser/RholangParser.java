package org.rholang.lang.parser;

import com.intellij.lang.ASTNode;
import com.intellij.lang.PsiBuilder;
import com.intellij.lang.PsiParser;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.psi.tree.IElementType;
import org.jetbrains.annotations.NotNull;

public class RholangParser implements PsiParser {
    private static final Logger LOG = Logger.getInstance(RholangParser.class);

    @NotNull
    @Override
    public ASTNode parse(@NotNull IElementType root, @NotNull PsiBuilder builder) {
        PsiBuilder.Marker rootMarker = builder.mark();

        // Get the full text from the builder
        String text = builder.getOriginalText().toString();

        // Check if the code is valid using the Rust parser
        boolean isValid = RholangParserCli.isValid(text);
        LOG.debug("Rholang code is valid: " + isValid);

        // For now, just consume all tokens
        // In a more complete implementation, we would parse the tree returned by the Rust parser
        // and build a corresponding PSI tree
        while (!builder.eof()) {
            builder.advanceLexer();
        }

        rootMarker.done(root);
        return builder.getTreeBuilt();
    }
}
