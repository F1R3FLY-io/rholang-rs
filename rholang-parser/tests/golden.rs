use rholang_parser::RholangParser;
use rstest::rstest;
use std::fs;
use std::path::PathBuf;

#[rstest]
fn golden_test(
    #[base_dir = "tests/corpus/"]
    #[files("*.rho")]
    path: PathBuf,
) {
    let mut settings = insta::Settings::new();
    settings.set_snapshot_path("corpus/golden_snapshots");

    settings.bind(|| {
        let name = path.file_stem().unwrap().to_string_lossy().to_string();

        let parser = RholangParser::new();
        let input = fs::read_to_string(path).expect("Failed to read input file");
        let result = parser.parse(input.as_str());

        // Each snapshot is named after the input file
        insta::assert_debug_snapshot!(name, result);
    })
}
