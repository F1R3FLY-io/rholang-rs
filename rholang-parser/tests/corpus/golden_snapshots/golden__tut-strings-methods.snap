---
source: rholang-parser/tests/golden.rs
expression: result
---
Good(
    [
        AnnProc {
            proc: New {
                decls: [
                    NameDecl {
                        id: Id {
                            name: "stdout",
                            pos: SourcePos {
                                line: 1,
                                col: 5,
                            },
                        },
                        uri: Some(
                            Uri(
                                "rho:io:stdout",
                            ),
                        ),
                    },
                ],
                proc: AnnProc {
                    proc: Par {
                        left: AnnProc {
                            proc: Par {
                                left: AnnProc {
                                    proc: Par {
                                        left: AnnProc {
                                            proc: Par {
                                                left: AnnProc {
                                                    proc: Send {
                                                        channel: NameVar(
                                                            Id(
                                                                Id {
                                                                    name: "stdout",
                                                                    pos: SourcePos {
                                                                        line: 4,
                                                                        col: 3,
                                                                    },
                                                                },
                                                            ),
                                                        ),
                                                        send_type: Single,
                                                        inputs: [
                                                            AnnProc {
                                                                proc: Method {
                                                                    receiver: AnnProc {
                                                                        proc: StringLiteral(
                                                                            "Hello World",
                                                                        ),
                                                                        span: SourceSpan {
                                                                            start: SourcePos {
                                                                                line: 4,
                                                                                col: 11,
                                                                            },
                                                                            end: SourcePos {
                                                                                line: 4,
                                                                                col: 24,
                                                                            },
                                                                        },
                                                                    },
                                                                    name: Id {
                                                                        name: "length",
                                                                        pos: SourcePos {
                                                                            line: 4,
                                                                            col: 25,
                                                                        },
                                                                    },
                                                                    args: [],
                                                                },
                                                                span: SourceSpan {
                                                                    start: SourcePos {
                                                                        line: 4,
                                                                        col: 11,
                                                                    },
                                                                    end: SourcePos {
                                                                        line: 4,
                                                                        col: 33,
                                                                    },
                                                                },
                                                            },
                                                        ],
                                                    },
                                                    span: SourceSpan {
                                                        start: SourcePos {
                                                            line: 4,
                                                            col: 3,
                                                        },
                                                        end: SourcePos {
                                                            line: 4,
                                                            col: 34,
                                                        },
                                                    },
                                                },
                                                right: AnnProc {
                                                    proc: Send {
                                                        channel: NameVar(
                                                            Id(
                                                                Id {
                                                                    name: "stdout",
                                                                    pos: SourcePos {
                                                                        line: 7,
                                                                        col: 3,
                                                                    },
                                                                },
                                                            ),
                                                        ),
                                                        send_type: Single,
                                                        inputs: [
                                                            AnnProc {
                                                                proc: Method {
                                                                    receiver: AnnProc {
                                                                        proc: StringLiteral(
                                                                            "Hello World",
                                                                        ),
                                                                        span: SourceSpan {
                                                                            start: SourcePos {
                                                                                line: 7,
                                                                                col: 11,
                                                                            },
                                                                            end: SourcePos {
                                                                                line: 7,
                                                                                col: 24,
                                                                            },
                                                                        },
                                                                    },
                                                                    name: Id {
                                                                        name: "slice",
                                                                        pos: SourcePos {
                                                                            line: 7,
                                                                            col: 25,
                                                                        },
                                                                    },
                                                                    args: [
                                                                        AnnProc {
                                                                            proc: LongLiteral(
                                                                                2,
                                                                            ),
                                                                            span: SourceSpan {
                                                                                start: SourcePos {
                                                                                    line: 7,
                                                                                    col: 31,
                                                                                },
                                                                                end: SourcePos {
                                                                                    line: 7,
                                                                                    col: 32,
                                                                                },
                                                                            },
                                                                        },
                                                                        AnnProc {
                                                                            proc: LongLiteral(
                                                                                8,
                                                                            ),
                                                                            span: SourceSpan {
                                                                                start: SourcePos {
                                                                                    line: 7,
                                                                                    col: 34,
                                                                                },
                                                                                end: SourcePos {
                                                                                    line: 7,
                                                                                    col: 35,
                                                                                },
                                                                            },
                                                                        },
                                                                    ],
                                                                },
                                                                span: SourceSpan {
                                                                    start: SourcePos {
                                                                        line: 7,
                                                                        col: 11,
                                                                    },
                                                                    end: SourcePos {
                                                                        line: 7,
                                                                        col: 36,
                                                                    },
                                                                },
                                                            },
                                                        ],
                                                    },
                                                    span: SourceSpan {
                                                        start: SourcePos {
                                                            line: 7,
                                                            col: 3,
                                                        },
                                                        end: SourcePos {
                                                            line: 7,
                                                            col: 37,
                                                        },
                                                    },
                                                },
                                            },
                                            span: SourceSpan {
                                                start: SourcePos {
                                                    line: 4,
                                                    col: 3,
                                                },
                                                end: SourcePos {
                                                    line: 7,
                                                    col: 37,
                                                },
                                            },
                                        },
                                        right: AnnProc {
                                            proc: Send {
                                                channel: NameVar(
                                                    Id(
                                                        Id {
                                                            name: "stdout",
                                                            pos: SourcePos {
                                                                line: 10,
                                                                col: 3,
                                                            },
                                                        },
                                                    ),
                                                ),
                                                send_type: Single,
                                                inputs: [
                                                    AnnProc {
                                                        proc: BinaryExp {
                                                            op: Concat,
                                                            left: AnnProc {
                                                                proc: StringLiteral(
                                                                    "Hello ",
                                                                ),
                                                                span: SourceSpan {
                                                                    start: SourcePos {
                                                                        line: 10,
                                                                        col: 11,
                                                                    },
                                                                    end: SourcePos {
                                                                        line: 10,
                                                                        col: 19,
                                                                    },
                                                                },
                                                            },
                                                            right: AnnProc {
                                                                proc: StringLiteral(
                                                                    "World",
                                                                ),
                                                                span: SourceSpan {
                                                                    start: SourcePos {
                                                                        line: 10,
                                                                        col: 23,
                                                                    },
                                                                    end: SourcePos {
                                                                        line: 10,
                                                                        col: 30,
                                                                    },
                                                                },
                                                            },
                                                        },
                                                        span: SourceSpan {
                                                            start: SourcePos {
                                                                line: 10,
                                                                col: 11,
                                                            },
                                                            end: SourcePos {
                                                                line: 10,
                                                                col: 30,
                                                            },
                                                        },
                                                    },
                                                ],
                                            },
                                            span: SourceSpan {
                                                start: SourcePos {
                                                    line: 10,
                                                    col: 3,
                                                },
                                                end: SourcePos {
                                                    line: 10,
                                                    col: 31,
                                                },
                                            },
                                        },
                                    },
                                    span: SourceSpan {
                                        start: SourcePos {
                                            line: 4,
                                            col: 3,
                                        },
                                        end: SourcePos {
                                            line: 10,
                                            col: 31,
                                        },
                                    },
                                },
                                right: AnnProc {
                                    proc: Send {
                                        channel: NameVar(
                                            Id(
                                                Id {
                                                    name: "stdout",
                                                    pos: SourcePos {
                                                        line: 13,
                                                        col: 3,
                                                    },
                                                },
                                            ),
                                        ),
                                        send_type: Single,
                                        inputs: [
                                            AnnProc {
                                                proc: Method {
                                                    receiver: AnnProc {
                                                        proc: StringLiteral(
                                                            "A40BE5",
                                                        ),
                                                        span: SourceSpan {
                                                            start: SourcePos {
                                                                line: 13,
                                                                col: 11,
                                                            },
                                                            end: SourcePos {
                                                                line: 13,
                                                                col: 19,
                                                            },
                                                        },
                                                    },
                                                    name: Id {
                                                        name: "hexToBytes",
                                                        pos: SourcePos {
                                                            line: 13,
                                                            col: 20,
                                                        },
                                                    },
                                                    args: [],
                                                },
                                                span: SourceSpan {
                                                    start: SourcePos {
                                                        line: 13,
                                                        col: 11,
                                                    },
                                                    end: SourcePos {
                                                        line: 13,
                                                        col: 32,
                                                    },
                                                },
                                            },
                                        ],
                                    },
                                    span: SourceSpan {
                                        start: SourcePos {
                                            line: 13,
                                            col: 3,
                                        },
                                        end: SourcePos {
                                            line: 13,
                                            col: 33,
                                        },
                                    },
                                },
                            },
                            span: SourceSpan {
                                start: SourcePos {
                                    line: 4,
                                    col: 3,
                                },
                                end: SourcePos {
                                    line: 13,
                                    col: 33,
                                },
                            },
                        },
                        right: AnnProc {
                            proc: Send {
                                channel: NameVar(
                                    Id(
                                        Id {
                                            name: "stdout",
                                            pos: SourcePos {
                                                line: 16,
                                                col: 3,
                                            },
                                        },
                                    ),
                                ),
                                send_type: Single,
                                inputs: [
                                    AnnProc {
                                        proc: Method {
                                            receiver: AnnProc {
                                                proc: StringLiteral(
                                                    "Hello World",
                                                ),
                                                span: SourceSpan {
                                                    start: SourcePos {
                                                        line: 16,
                                                        col: 11,
                                                    },
                                                    end: SourcePos {
                                                        line: 16,
                                                        col: 24,
                                                    },
                                                },
                                            },
                                            name: Id {
                                                name: "hexToBytes",
                                                pos: SourcePos {
                                                    line: 16,
                                                    col: 25,
                                                },
                                            },
                                            args: [],
                                        },
                                        span: SourceSpan {
                                            start: SourcePos {
                                                line: 16,
                                                col: 11,
                                            },
                                            end: SourcePos {
                                                line: 16,
                                                col: 37,
                                            },
                                        },
                                    },
                                ],
                            },
                            span: SourceSpan {
                                start: SourcePos {
                                    line: 16,
                                    col: 3,
                                },
                                end: SourcePos {
                                    line: 16,
                                    col: 38,
                                },
                            },
                        },
                    },
                    span: SourceSpan {
                        start: SourcePos {
                            line: 4,
                            col: 3,
                        },
                        end: SourcePos {
                            line: 16,
                            col: 38,
                        },
                    },
                },
            },
            span: SourceSpan {
                start: SourcePos {
                    line: 1,
                    col: 1,
                },
                end: SourcePos {
                    line: 17,
                    col: 2,
                },
            },
        },
    ],
)
